packages/core/src/shader/Shader.ts

```diff
@@ -123,17 +137,25 @@ export class Shader {
   /** The name of shader. */
   readonly name: string;
 
-  /** @internal */
-  _shaderId: number = 0;
+  /**
+   *  Shader passes.
+   */
+  get passes(): ReadonlyArray<ShaderPass> {
+    return this._passes;
+  }
 
-  private _vertexSource: string;
-  private _fragmentSource: string;
+  private _passes: ShaderPass[] = [];
 
-  private constructor(name: string, vertexSource: string, fragmentSource: string) {
-    this._shaderId = Shader._shaderCounter++;
+  private constructor(name: string, vertexSourceOrShaderPasses: string | ShaderPass[], fragmentSource?: string) {
     this.name = name;
-    this._vertexSource = vertexSource;
-    this._fragmentSource = fragmentSource;
+
+    if (typeof vertexSourceOrShaderPasses === "string") {
+      this._passes.push(new ShaderPass(vertexSourceOrShaderPasses, fragmentSource));
+    } else {
+      for (let i = 0, n = vertexSourceOrShaderPasses.length; i < n; i++) {
+        this._passes.push(vertexSourceOrShaderPasses[i]);
+      }
```

Comment:
If vertexSourceOrShaderPasses.length is 0, will not has default pass

Reply:
done